src/vcpkg.cpp

```diff
@@ -156,6 +157,7 @@ namespace vcpkg::Checks
     // Implements link seam from basic_checks.h
     void on_final_cleanup_and_exit()
     {
+        BinaryCache::wait_for_async_complete();
```

Comment:
I have changed the `BinaryCache::wait_for_async_complete()` implementation so it does not deadlock anymore. 

I also moved the call to `Checks::exit_with_code` which is not called when crtl+c is handled. (I personally would like to have a way to terminate vcpkg but wait until the binary cache is done so that I don't lose progress.) 

And I prefer it when build packages are uploaded to the binary caches before vcpkg exits because of an error, otherwise I have to build the already build packages again at a later point when there is no cache entry.

Reply:
Agreed that it is, desirable to finish uploading on "understood" errors. For example, if a package failed to build or failed to be installed.

I was also wrong about my original assessment of a deadlock. My concern was the call path of the binary upload thread calling `Checks::unreachable()` or `.value_or_exit()`, but it seems that `std::thread::join()` does have a carve-out to handle this specific case: it will throw a `resource_deadlock_would_occur` if you try to join yourself.

I've put some other concerns below, but I don't want those to distract from my main point: I think the best approach for vcpkg right now is to add calls from `Install::perform()` etc to `BinaryCache::wait_for_async_complete()` before any "user-facing" error, such as the exit guarded by `result.code != BuildResult::SUCCEEDED && keep_going == KeepGoing::NO`. This is motivated by the perspective that it's always safer to terminate than to join and possibly deadlock / race condition / etc. We must make it as trivial / correct-by-construction as possible to guarantee that the binary cache thread NEVER attempts to wait on itself.

---

There's still a UB data race if the main thread and binary upload thread attempt to exit at the same time:
> Concurrently calling join() on the same thread object from multiple threads constitutes a data race that results in undefined behavior.
> -- https://en.cppreference.com/w/cpp/thread/thread/join

There's also a serious "scalability" problem if we ever want a second background thread for whatever reason, because BGThread A would join on BGThread B, while BGThread B tries to join on BGThread A. This might be solvable with ever more complex structures, such as a thread ownership DAG that gets threads to join only on their direct children, but I don't think the benefit is worth the cost.

Reply:
The UB and the joining itself could simply be prevented by doing a `if (std::this_thread::get_id() == instance->push_thread.get_id())`. My concern with the explicit approach is that it is easy to forget to call the waiting function of the BinaryCache and every time you want to exit you have to remember to call it. This seems to me to be very prone to human error.

Reply:
I have now implemented that (I hope I have not missed a location, I am not 100% sure)

Reply:
@ras0219-msft Is there anything left that is preventing this PR from being merged?